### PR TITLE
Update to Rails 5

### PR DESCRIPTION
**Pull request by aitbw**
References
==========
* Related issue: #784, #2627
* Related PR: AyuntamientoMadrid/consul#1458

Objectives
==========
* Update CONSUL to use Rails 5 (and Ruby 2.4.1) :tada: 

Visual Changes (if any)
=======================
* None

Notes
=====================
* This is a **major** update, so be careful when updating your current CONSUL installation
* Considering reading the official [Ruby on Rails guide](https://guides.rubyonrails.org/v5.0/upgrading_ruby_on_rails.html#upgrading-from-rails-4-2-to-rails-5-0) when upgrading your CONSUL installation with this PR

### CODE REVIEW

**Comment by aitbw** [comment]
We found several errors when deploying this to a staging development (as shown [here](https://github.com/AyuntamientoMadrid/consul/pull/1458#issuecomment-393902842)), so we decided to close this so we can rework this migration from scratch

**Comment by javierm** [comment]
@aitbw Thank you so much for this one! :tada:

One small question: is it ready for review, or is it still work in progress?

**Comment by aitbw** [comment]
@javierm technically yes because all deprecation warnings have been solved, Travis is all green, dependencies updates have been met and everything seems to be working as expected 

That said, I'm currently working on deploying a CONSUL instance to Heroku to see how it fares on a production environment —that's because when we deployed this to Madrid (via AyuntamientoMadrid#1458), several errors and exceptions popped up on the `staging` environment

Our principal suspect of this behavior is the yet uncompleted `Spending Proposal to Budget Investment` migration but we also have our eyes on the Ruby update as well (from 2.3.2 to 2.4.1), although the latter is less likely (but not totally discarded)

**Comment by javierm** [comment]
@aitbw Sounds great! I'll have a look at the changes then.

File: spec/lib/census_api_spec.rb [hunk 1]
```diff
@@ -1,7 +1,7 @@
 require 'rails_helper'
 
-describe CensusApi do
-  let(:api) { described_class.new }
+describe 'CensusApi' do
```
Comment (javierm) [inline spec/lib/census_api_spec.rb]: I don't understand the reason for commit 424dfdf5. Any ideas?

File: Gemfile [hunk 2]
```diff
@@ -1,76 +1,77 @@
...
 gem 'omniauth', '~> 1.8.1'
 gem 'omniauth-facebook', '~> 4.0.0'
-gem 'omniauth-google-oauth2', '~> 0.4.0'
+gem 'omniauth-google-oauth2', '~> 0.4.1'
 gem 'omniauth-twitter', '~> 1.4.0'
 gem 'paperclip', '~> 5.2.1'
-gem 'paranoia', '~> 2.4.0'
-gem 'pg', '~> 0.21.0'
-gem 'pg_search', '~> 2.0.1'
+gem 'paranoia', '~> 2.3.1'
```
[the remark this answers is not in the document]
Comment (aitbw) [reply Gemfile]: You're right, @javierm —we definitely downgraded `paranoia` (see #2253) but I can't find an upgrade for `pg` to version 1.0.0 (its latest update came from #1820)

File: app/controllers/admin/spending_proposals_controller.rb [hunk 3]
```diff
@@ -46,6 +47,22 @@ def spending_proposal_params
...
+      @filter_params ||= filter_params
+    end
+
+    def filter_params
+      params.permit(:geozone_id, :administrator_id, :tag_name, :valuator_id)
+    end
+
+    def load_filter_params
+      @filter_params ||= filter_params
+    end
```
[the remark this answers is not in the document]
Comment (aitbw) [reply app/controllers/admin/spending_proposals_controller.rb]: Oops! Removing them asap, thanks! :raised_hands:

File: Gemfile (hunk 2 above)
[the remark this answers is not in the document]
Comment (javierm) [reply Gemfile]: Maybe we can leave `pg` at `0.21.0` instead of downgrading it to `0.20.0`?

[the remark this answers is not in the document]
Comment (aitbw) [reply Gemfile]: Ohhhh I haven't noticed we had downgraded that one too! Fixing it as we speak

File: app/models/poll/null_result.rb [hunk 4]
```diff
@@ -0,0 +1,23 @@
+class Poll::NullResult < ApplicationRecord
```
[the remark this answers is not in the document]
Comment (aitbw) [reply app/models/poll/null_result.rb]: Maybe it got re-introduced due to conflict resolution —good catch! Already fixed, I'm waiting for this Travis build to complete to push it

File: app/models/poll/final_recount.rb [hunk 5]
```diff
@@ -0,0 +1,19 @@
+class Poll
```
[the remark this answers is not in the document]
Comment (aitbw) [reply app/models/poll/final_recount.rb]: Maybe it got re-introduced due to conflict resolution —good catch! Already fixed, I'm waiting for this Travis build to complete to push it

File: app/models/poll/white_result.rb [hunk 6]
```diff
@@ -0,0 +1,23 @@
+class Poll::WhiteResult < ApplicationRecord
```
[the remark this answers is not in the document]
Comment (aitbw) [reply app/models/poll/white_result.rb]: Maybe it got re-introduced due to conflict resolution —good catch! Already fixed, I'm waiting for this Travis build to complete to push it

File: app/models/poll/recount.rb [hunk 7]
```diff
@@ -1,36 +1,43 @@
-class Poll::Recount < ActiveRecord::Base
```
[the remark this answers is not in the document]
Comment (javierm) [reply app/models/poll/recount.rb]: I've just realized we're also changing inheritance (replacing `ActiveRecord::Base` with `ApplicationRecord`). So the changes are fine with me.

File: spec/lib/wysiwyg_sanitizer_spec.rb [hunk 8]
```diff
@@ -1,8 +1,8 @@
 require 'rails_helper'
 
-describe WYSIWYGSanitizer do
+describe 'WYSIWYGSanitizer' do
```
Comment (javierm) [inline spec/lib/wysiwyg_sanitizer_spec.rb]: It looks like the changes using strings introduced in 424dfdf were reverted later. To keep it consistent, what do you think about changing the first line back to `describe WYSIWYGSanitizer do`?

**Comment by javierm** [comment]
I've downloaded the branch. On my machine most things seem to be working like a charm! Only a couple of gotchas so far. The first time I executed a rake task, I got the message:

```
rake aborted!
ActiveRecord::NoEnvironmentInSchemaError:

Environment data not found in the schema. To resolve this issue, run:

        bin/rails db:environment:set RAILS_ENV=development
```
After executing that command, everything works fine. @aitbw Do you think it should be documented in the README (or somewhere else)?

And, when dropping the database and running `rake db:migrate`, I got the message:

```
StandardError: An error has occurred, this and all later migrations canceled:

PG::SyntaxError: ERROR:  syntax error at or near "/"
LINE 1: ...gislation_proposals_count" ON "tags"  (legislation/proposals...
                                                             ^
: CREATE  INDEX  "index_tags_on_legislation_proposals_count" ON "tags"  (legislation/proposals_count)
/home/cjt/dev/consul/db/migrate/20170915101519_add_legislation_proposals_count_to_tags.rb:4:in `change'

Caused by:
ActiveRecord::StatementInvalid: PG::SyntaxError: ERROR:  syntax error at or near "/"
LINE 1: ...gislation_proposals_count" ON "tags"  (legislation/proposals...
                                                             ^
: CREATE  INDEX  "index_tags_on_legislation_proposals_count" ON "tags"  (legislation/proposals_count)
/home/cjt/dev/consul/db/migrate/20170915101519_add_legislation_proposals_count_to_tags.rb:4:in `change'

Caused by:
PG::SyntaxError: ERROR:  syntax error at or near "/"
LINE 1: ...gislation_proposals_count" ON "tags"  (legislation/proposals...
                                                             ^
/home/cjt/dev/consul/db/migrate/20170915101519_add_legislation_proposals_count_to_tags.rb:4:in `change'
Tasks: TOP => db:migrate
(See full trace by running task with --trace)
```

@aitbw I'll have a look at this error, but if you have any ideas, they're welcome!

**Comment by javierm** [comment]
@aitbw It looks like we need to change a couple of migration files. Here's the patch:

```
diff --git a/db/migrate/20170915101519_add_legislation_proposals_count_to_tags.rb b/db/migrate/20170915101519_add_legislation_proposals_count_to_tags.rb
index 9ce6ace..8c47394 100644
--- a/db/migrate/20170915101519_add_legislation_proposals_count_to_tags.rb
+++ b/db/migrate/20170915101519_add_legislation_proposals_count_to_tags.rb
@@ -1,6 +1,6 @@
 class AddLegislationProposalsCountToTags < ActiveRecord::Migration
   def change
     add_column :tags, "legislation/proposals_count", :integer, default: 0
-    add_index :tags, "legislation/proposals_count"
+    add_index :tags, :"legislation/proposals_count"
   end
 end
diff --git a/db/migrate/20170922101247_add_legislation_processes_count_to_tags.rb b/db/migrate/20170922101247_add_legislation_processes_count_to_tags.rb
index 698a66a..4ae67b7 100644
--- a/db/migrate/20170922101247_add_legislation_processes_count_to_tags.rb
+++ b/db/migrate/20170922101247_add_legislation_processes_count_to_tags.rb
@@ -1,6 +1,6 @@
 class AddLegislationProcessesCountToTags < ActiveRecord::Migration
   def change
     add_column :tags, "legislation/processes_count", :integer, default: 0
-    add_index :tags, "legislation/processes_count"
+    add_index :tags, :"legislation/processes_count"
   end
 end
```

**Comment by javierm** [comment]
@aitbw It would also be great to regenerate `db/schema.rb` (by running `rake db:migrate` locally) and include it as part of this pull request, since the format of that file has changed slightly. Otherwise we would have many changes in that file every time we add a new migration. Does it make sense to you?

**Comment by aitbw** [comment]
Just pushed a commit with the regenerated schema, @javierm —that should fix those issues :tada:

File: spec/lib/census_api_spec.rb (hunk 1 above)
Comment (aitbw) [reply spec/lib/census_api_spec.rb]: Sounds good to me! :ok_hand:

**Comment by javierm** [comment]
@aitbw Great! I think we still need to change the migration files, though. Running `rake db:drop db:create db:migrate` still gives me the same error until I apply the patch I mentioned.

**Comment by aitbw** [comment]
Hi @javierm, I had to rewrite Git history to remove one commit because I thought it was breaking our build (it was not, the problem came was coming from the Travis side) so make sure (if you want to keep your local branch up to date) to execute `git pull origin <my-branch> --rebase` before

That said, I think I found a better solution for the database migration problem —apparently, it has to do with `database_cleaner` screwing up the `ar_internal_metadata` table, so we need to fix our `spec_helper` file with the following:

```ruby
DatabaseCleaner.clean_with(
  :truncation,
  except: %w(ar_internal_metadata)
)
```

(Thanks once again, [StackOverflow](https://stackoverflow.com/a/38209363/4844313)! :tada:)

Let me know how it fares and if it does the trick, I'll upload a patch

**Comment by javierm** [comment]
Hi, @aitbw. Thanks for all these changes! From my side, the code review is done, and now we "only" need to make sure everything works properly.

I've applied the changes and I'm still getting the error in the migrations. By the way, I'm getting the error in the development environment, so `database_cleaner` might not be related.

I'd like to make sure it isn't an issue with my machine. What happens on your machine if you do the following?

1. Backup your development database (if you'd like to keep it)
2. Run `rake db:drop`
3. Run `rake db:create`
4. Run `rake db:migrate`

On my machine I'm getting an error when applying the migration in `/home/cjt/dev/consul/db/migrate/20170915101519_add_legislation_proposals_count_to_tags.rb`

File: Gemfile [hunk 9]
```diff
@@ -80,20 +81,21 @@ group :test do
...
+  gem 'selenium-webdriver', '~> 3.11.0', require: false
 end
 
 group :development do
   gem 'capistrano', '~> 3.10.1', require: false
+  gem 'capistrano3-delayed-job', '~> 1.7.3'
   gem 'capistrano-bundler', '~> 1.2', require: false
   gem 'capistrano-rails', '~> 1.3.1', require: false
-  gem 'capistrano3-delayed-job', '~> 1.7.3'
+  gem 'fuubar', '~> 2.3.1'
```
Comment (javierm) [inline Gemfile]: While working on #1611 isn't in the `Gemfile` anymore in the `master` branch. Does it make sense to remove this line from this pull request?

**Comment by aitbw** [comment]
:thinking: It failed but you shouldn't use `rake db:migrate` to setup the database as it will run **all** of the migrations under `db/migrate`, which is error prone and not suggested. Can you try with `rake db:reset`, @javierm?

File: Gemfile (hunk 9 above)
Comment (aitbw) [reply Gemfile]: It's not even on Consul's `master` branch :sweat_smile: and I don't see it being used anywhere so let's remove it for the time being :ok_hand:

**Comment by aitbw** [comment]
I think we could edit the docs accordingly to use `rake db:reset` or `rake db:setup` instead of `rake db:migrate` (or we could also enforce `rails db:setup` usage (according to [this](https://guides.rubyonrails.org/v5.0/upgrading_ruby_on_rails.html#use-bin-rails-for-running-tasks-and-tests)))

I do understand we could easily fix the conflicting files to avoid such errors (even when the output is **exactly** the same as of now) but I don't feel comfortable modifying already-commited-and-ran migrations —especially since they have already been applied on production environments

**Comment by javierm** [comment]
Ideally we would fix the migrations without editing them. I've done some research and come up with nothing so far. No clear solution since this was such an unexpected issue (who would have thought upgrading to Rails 5 would break existing migration files). And leaving broken migrations which will generate an error if being run by forks who haven't already run them doesn't look like a good option either :thinking:.

So suggestions are welcome!

P.S. Regarding updating the documentation, I guess that's something only @MariaCheca and Rai can decide.

**Comment by javierm** [comment]
@aitbw BTW, so I understand potentials issues better, are there any specific reasons why you don't feel comfortable editing the syntax of migrations already run on production? Sorry if I'm suggesting editing migrations while there are some dangers I'm not aware of :sweat_smile:!

**Comment by aitbw** [comment]
>And leaving broken migrations which will generate an error if being run by forks who haven't already run them doesn't look like a good option either

Haven't thought of that, tbh —which makes a pretty good case for that patch (because you wouldn't expect a DB reset from scratch on a prod. machine)

>are there any specific reasons why you don't feel comfortable editing the syntax of migrations already run on production?

Alongside potential issues that may arise, it's an [antipattern](https://guides.rubyonrails.org/v5.0/active_record_migrations.html#changing-existing-migrations) and whilst there are some rules I might be willing to break, this is one of those that don't seem reasonable (at least for me) —in addition to being already commited to source control and already ran on prod. environments, take into consideration the amount of forks that we have and how this could affect them —even for such a small patch

Don't get me wrong, tho: I'm all in to solve this particular issue —I'm just trying to comprehend how this would affect all involved parties

**Comment by javierm** [comment]
@aitbw No worries! Obviously we're both trying to find a solution which works for everyone :wink:.

When I read the documentation I understood the antipattern only refers to changes affecting the database structure and/or contents, since it also says:

> In general, editing existing migrations is not a good idea. (...) Instead, you should write a new migration that performs the changes you require.

And writing a new migration wouldn't solve anything in this case.

That's just my interpretation, though; I could be wrong.

> in addition to being already commited to source control and already ran on prod. environments, take into consideration the amount of forks that we have and how this could affect them

That's exactly the point I'd like to know about! How could the patch affect those forks or the production environment? I thought it wouldn't affect them at all :thinking:.

P.S. We're not in touch with any Rails core members who can clarify what to do in these cases, are we?

**Comment by mlovic** [comment]
Is there a reason we're using Ruby version 2.4.1, and not a later version? (2.4.4 is the latest 2.4.x version)

Among other things, it would allow us to use the official Docker Ruby image for `2.4.4`, which is built on Debian 9 ("stretch") and has up-to-date packages. This would allow us to upgrade chromium from v57 to v67, which would:

1. Fix order-dependency of JS specs (#2849)

2. Allow us to use the latest version of chromedriver (#2818)

**Comment by javierm** [comment]
@aitbw I'm trying to deploy this branch to my server, just to have a look at possible production issues. I'm not using the official installer because I don't use the root user on my server, I use rbenv, and I wanted more control so I didn't break anything in the process.

So far, I've run into these issues:

1. Capistrano doesn't execute `rake db:create` before executing `rake assets:precompile`, resulting in an error. Manually running `rake db:create` on the server solved the problem. I don't think this point is related to Rails 5 at all, so we can probably ignore it in this PR.
2. The migrations mentioned on this PR failed. As a temporary solution, I applied the mentioned patch to this branch on my fork.
3. The task `deploy:restart` fails with the message `no such file or directory: /etc/init.d/unicorn_consul`.

I'm stuck in the last point. Do you know what the contents of `unicorn_consul` are supposed to be like?

**Comment by javierm** [comment]
@aitbw That's alright! Just a few minutes ago I finally made my server work, after a lot of trial and error with the unicorn script. I also had to configure nginx to play nice with Rails `force_ssl` configuration option, but again I don't think it has anything to do with upgrading to Rails 5.

So everything seems to be OK with this PR :tada:. Thank you so much for everything you've done! Now we only need to wait for everyone to come back and do some heavy testing. I guess there isn't a place to check which exceptions we got two months ago (mentioned in https://github.com/AyuntamientoMadrid/consul/pull/1458#issuecomment-393902842), is there?

@MariaCheca Is there a way to get a sample `unicorn_consul` script? I've already got a sample `unicorn.rb` configuration file, since it's present in our installer.

**Comment by javierm** [comment]
@aitbw It looks like, probably due to changes in Capybara/Selenium/chromedriver, we might have new flakies. In a [Travis build on this branch](https://travis-ci.org/javierm/consul/jobs/417194624) there was a failure I hadn't seen before. It's hard to explain, so I'll try my best.

Here's the test:
```ruby
within("#budget_investment_#{investment1.id}") do
  find('.remove a').click
end

visit budget_investments_path(budget, heading_id: new_york.id)
```
The first click generates an AJAX request, and the second one generates a normal request. What happens if the AJAX request takes too long? We can emulate it by changing the code in `app/controllers/budgets/ballot/lines_controller.rb`:
```ruby
def destroy
  sleep 5 # <==== New line. On my machine, sleep 1 produces the same effect.
  @investment = @line.investment
  load_heading

  @line.destroy
  load_investments
end
```
With this change, the test always fails on the `rails5` branch, and always passes on `master`. I only got the same error message as the build once, though :confused:. Updating `chromedriver` doesn't seem to solve the problem.

I'm not sure whether we should investigate it. IMHO, it's just a thing to take into account: after the upgrade we might have to change a few tests doing AJAX requests.

**Comment by aitbw** [comment]
@javierm Yeah, I'm aware of that flaky but it's not related to this PR :confused: it's been showing up from time to time but I haven't got time to investigate it but in case it is, we still haven't updated our `test` gems to their latest versions :sweat_smile: so we might as well get on it to see how it fares

**Comment by javierm** [comment]
> Yeah, I'm aware of that flaky but it's not related to this PR

You mean it fails on master too? If so, you're right. I thought it only failed on this branch because it's the only branch where I could reproduce it.

**Comment by voodoorai2000** [comment]
Closed in favor of https://github.com/consul/consul/pull/3414

**Pull request closed without merging**